be/src/vec/olap/block_reader.cpp

```diff
@@ -133,6 +133,29 @@ Status BlockReader::init(const ReaderParams& read_params) {
         return status;
     }
 
+    // only group by key num <= 3 and has bitmap_union col, enable block aggregate
+    if (config::enable_block_aggregate_in_scanner && _aggregation &&
+        (_tablet->keys_type() == AGG_KEYS)) {
+        int agg_key_size = 0;
+        for (int i = 0; i < return_column_size; ++i) {
+            auto cid = read_params.origin_return_columns->at(i);
+            auto& col = _tablet->tablet_schema().column(cid);
+            if (col.is_key() && ++agg_key_size > 3) {
```

Comment:
if `col1` and `col2` are both table keys and agg keys, `_block_aggregation` should also be false because we will get a block  which is already aggregated

Reply:
no. If `col1` and `col2` are both table keys and agg keys, the block we get is aggregated based on all keys in the aggregation table, not based on the group by key column. So `_block_aggregation` still needs to be true